src/fswatch_win/fswatch_win_stubs.c

```diff
@@ -0,0 +1,305 @@
+  HANDLE signal;
+  struct wenv *wenv;
+};
+
+struct wenv {
+  struct wenv *next;
+  char *dirpath;
+  HANDLE dir_handle;
+  OVERLAPPED *overlapped;
+  char *buffer;
+  struct fsenv *fsenv;
+};
+
+static void push_info(struct wenv *wenv) {
+  struct events *ev = malloc(sizeof(struct events));
+  struct fsenv *fsenv = wenv->fsenv;
+  ev->info = (FILE_NOTIFY_INFORMATION *)wenv->buffer;
+  ev->dirpath = wenv->dirpath;
+
+  /* Set fsenv->events = ev atomically */
+  do {
+    ev->next = fsenv->events;
+  } while (InterlockedCompareExchangePointer((void **)&(fsenv->events), ev, ev->next) != ev->next);
+
+  SetEvent(fsenv->signal); /* wakeup OCaml thread */
+}
+
+/* Retrieve the list of events from the shared list. */
+static struct events* pop_events(struct fsenv *fsenv) {
+  struct events *res;
+
+  /* Set fsenv->events = NULL atomically */
+  do {
+    res = fsenv->events;
+  } while(InterlockedCompareExchangePointer((void **)&(fsenv->events), NULL, res) != res);
+
+  return res;
+}
+
+/* Same order as Event.action */
+static DWORD actions[] = {
+  FILE_ACTION_ADDED,
+  FILE_ACTION_REMOVED,
+  FILE_ACTION_MODIFIED,
+  FILE_ACTION_RENAMED_OLD_NAME,
+  FILE_ACTION_RENAMED_NEW_NAME,
+};
+
+static value Val_action(DWORD action) {
+  for (int tag = 0; tag < sizeof(actions) / sizeof(DWORD); tag ++) {
+    if (actions[tag] == action)
+      return Val_int(tag);
+  }
+  CAMLassert(0);
+  return Val_unit; /* unreachable */
+}
+
+static value parse_events(struct events *events) {
+  CAMLparam0();
+  CAMLlocal5(res, cell, ev, v_directory, v_path);
+  FILE_NOTIFY_INFORMATION *info;
+  DWORD name_len;
+  WCHAR *file_name;
+
+  while (events != NULL) {
+    info = events->info;
+    v_directory = caml_copy_string(events->dirpath);
+
+    for (;;) {
+      name_len = info->FileNameLength / sizeof(WCHAR);
+      file_name = caml_stat_alloc(sizeof(WCHAR) * (name_len + 1));
+      memcpy(file_name, info->FileName, sizeof(WCHAR) * name_len);
+      file_name[name_len] = 0;
+      v_path = caml_copy_string_of_utf16(file_name);
+      caml_stat_free(file_name);
+
+      /* Allocate Event.t */
+      ev = caml_alloc_tuple(3);
+      Store_field(ev, 0, v_directory);
+      Store_field(ev, 1, v_path);
+      Store_field(ev, 2, Val_action(info->Action));
+
+      /* Allocate list cell */
+      cell = caml_alloc_tuple(2);
+      Store_field(cell, 0, ev);
+      Store_field(cell, 1, res);
+      res = cell;
+
+      if (info->NextEntryOffset == 0)
+        break;
+
+      info = (FILE_NOTIFY_INFORMATION *)((char*)info + info->NextEntryOffset);
+    }
+
+    caml_stat_free(events->info);
+    events = events->next;
+  }
+
+  CAMLreturn(res);
+}
+
+value fswatch_win_create(value v_unit) {
+  CAMLparam1(v_unit);
+  struct fsenv *fsenv = malloc(sizeof(struct fsenv));
+  HANDLE iocp, signal;
+  DWORD err;
+
+  if ((iocp = CreateIoCompletionPort(INVALID_HANDLE_VALUE, NULL, 0, 1)) == NULL) {
+    win32_maperr(GetLastError());
+    uerror("CreateIoCompletionPort", Nothing);
```

Comment:
IIRC, `uerror` does not exist in OCaml 5.0 anymore.

Reply:
Ah, right. As we don't test Windows under 5.0, this is not tested. Do you know if there is something that works both for 4.* and 5.* versions?

Reply:
not that I know of. I think that using `caml_uerror` + something like this would work:

```c
#if OCAML_VERSION_MAJOR < 5
#define caml_uerror uerror
#endif
```

Reply:
I would also define `CAML_NAME_SPACE` (it removes some short unprefixed `caml_` identifiers before OCaml 5.0):
```c
#if OCAML_VERSION_MAJOR < 5
#define CAML_NAME_SPACE
#define caml_uerror uerror
#endif
```

and also `WIN32_LEAN_AND_MEAN`.

Reply:
Thanks, fixed (both suggestions).